bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+    Protocol,
+    TypeVar,
+)
+class Member(NamedTuple):
+    id: str
+    name: str
+    email: str
+    type: str
+    access_all: Optional[bool]
+    # groups: Tuple[str, ...]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Member:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+        return Member(
+            id=data["member_id"],
+            name=data["member_name"],
+            email = data["email"],
+            type = data["type"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[member]]",
+            f"member_id = {self.id}",
+            f"member_name = {self.name}",
+            f"email = {self.email}",
+            f"type = {str(self.type)}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+class GroupMember(NamedTuple):
+    member_id: int
+    member_name: str
+    group_name: str
+
+    def get_id(self) -> str:
+        # Our generic differ has the ability to turn add/removes into changes
+        # for pairs with the same id, but this does not apply to memberships,
+        # which do not themselves have an id, so the identity to group on is
+        # the value itself.
+        return f"{self.member_id}@{self.group_name}"
+
+    def format_toml(self) -> str:
+        # Needed to satisfy Diffable, but not used in this case.
+        raise Exception(
+            "Group memberships are not expressed in toml, "
+            "please print the diffs in some other way."
+        )
+
+class Group(NamedTuple):
+    id: str
+    name: str
+    access_all: Optional[bool]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Group:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+
+        return Group(
+            id=data["group_id"],
+            name=data["group_name"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[group]]",
+            f"group_id = {self.id}",
+            f"group_name = {self.name}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+
+class MemberCollectionAccess(NamedTuple):
+    name: str
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> MemberCollectionAccess:
+        return MemberCollectionAccess(
+            name=data["member_name"],
+        )
+
+    def format_toml(self) -> str:
+        return (
+            "{ member_name = "
+            + self.name
+            + '"}'
+        )
+
+class GroupCollectionAccess(NamedTuple):
+    name: str
+    access: str
```

Comment:
I would make this an enum as well instead of a string.

Reply:
Same question as below, the API return `false|true` and our configuration file use `read|write`. How do we use our enum in this case ?

> In the context of from_toml_dict method for the Member class, we don't get the type from the configuration file as int but as a string to be user friendly with owner|user.... The values in MemberType are not matching, will not resolve the name and throw an error.

Reply:
You can do something like
```python
Access.write if api_response.can_write else Access.read_only
```
to convert the bool to the enum value.